py/desispec/io/raw.py

```diff
@@ -46,105 +69,130 @@ def read_raw(filename, camera, fibermapfile=None, **kwargs):
-                    hdu_header = fx[hdu].header
-                    for key in hdu_header:
-                        if ( key not in skipkeys ) and ( key not in header ) :
-                            log.debug("adding {} = {}".format(key,hdu_header[key]))
-                            header[key] = hdu_header[key]
-                        else :
-                            log.debug("key %s already in header or in skipkeys"%key)
-                else :
-                    log.warning("warning HDU %s not in fits file"%str(hdu))
-
-        kwargs.pop("fill_header")
+        inherited = True
+        log.info('Camera header %s will INHERIT from HDU 0.', camera)
+        header.extend(fx[0].header, strip=True, unique=True)
+
+    if fill_header is None:
+        if inherited:
+            hdus = []
+        else:
+            hdus = [0,]
+        if 'PLC' in fx:
+            hdus.append('PLC')
+    elif isinstance(fill_header, list):
+        hdus = fill_header
+    else:
+        msg = 'Unknown type for fill_header!'
+        log.critical(msg)
+        raise ValueError(msg)
+
+    log.info('Will add header keywords from HDUs %s.', str(hdus))
+
+    for hdu in hdus:
+        if hdu in fx or int(hdu) in fx:
+            header.extend(fx[hdu].header, strip=True, unique=True)
+        else:
+            log.warning("HDU %s is not in FITS file.", str(hdu))
 
     fx.close()
     duration = time.time() - t0
     log.info(iotime.format('read', filename, duration))
-
+    #
+    # Other cleanup of headers
+    #
+    longstrn = fits.Card('LONGSTRN', 'OGIP 1.0', 'The OGIP Long String Convention may be used.')
+    if 'MODULE' in primary_header:
+        log.debug("Inserting LONGSTRN keyword before MODULE.")
+        primary_header.insert('MODULE', longstrn)
+    else:
+        log.debug("Inserting LONGSTRN keyword before EXTNAME.")
+        primary_header.insert('EXTNAME', longstrn)
+    log.debug("Renaming EPOCH to EQUINOX in primary_header.")
+    try:
+        primary_header.rename_keyword('EPOCH', 'EQUINOX')
+    except ValueError:
+        # If EPOCH is not present at all.
+        pass
+    log.debug("Renaming EPOCH to EQUINOX in header.")
+    try:
+        header.rename_keyword('EPOCH', 'EQUINOX')
+    except ValueError:
+        pass
+    #
+    # Run preproc()
+    #
     img = desispec.preproc.preproc(rawimage, header, primary_header, **kwargs)
-
+    #
+    # Load fibermap data.
+    #
     if fibermapfile is not None and os.path.exists(fibermapfile):
         fibermap = desispec.io.read_fibermap(fibermapfile)
     else:
         log.warning('creating blank fibermap')
         fibermap = desispec.io.empty_fibermap(5000)
 
     #- Add image header keywords inherited from raw data to fibermap too
-    desispec.io.util.addkeys(fibermap.meta, img.meta)
+    # BAW: This is unnecssary, because a fibermap file constructed by
+    # assemble_fibermap() will already have the raw data keywords.
+    # desispec.io.util.addkeys(fibermap.meta, img.meta)
```

Comment:
I think this is necessary for DARK, ZERO, ARC, FLAT exposures that don't have fibermaps from assemble_fibermap.  `desispec.io.util.addkeys` will not overwrite a pre-existing keys (i.e. it only adds new ones) so I think this is harmless to call.  If that causes incorrect keywords to be added, those should be added to skipkeys or stdkeys in desispec.io.util.addkeys.  Or otherwise we need to catch the case where read_raw used `empty_fibermap` and needs to propagate keywords from the raw data.

More generally, we want the raw data headers propagated to the fibermap HDU, regardless of whether that comes via assemble_fibermap -> fibermap file, or whether that comes via empty_fibermap + adding raw data headers.

Reply:
OK, that's reasonable, but I want to make sure that there are only two possible channels:  *either* `assemble_fibermap()` *or* `empty_fibermap()`.  That is there is no other possible way to add a fibermap HDU.

Reply:
I believe those are the only two ways that a fibermap gets created.  But we should preserve the ability to create an empty_fibermap that isn't tied to a pre-existing raw data file (e.g. for simulations and testing), i.e. header propagation into empty_fibermap could be an option, but shouldn't be required.  Or otherwise it's fine with me if empty_fibermap doesn't pretend to add raw data headers and io.raw.read_raw adds them when needed.

Reply:
OK, that's what I've got in the latest version.  Please take a look again.